[kayabuttertoastt] iP

kayabuttertoastt opened the pull request:
# ChatBot 

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

## Features 
* Add Todo
* Add Deadlines 
* Add Events 
* Delete Tasks
* Mark/Unmark Tasks 

## Getting Started 
1. Download it from [here](https://nus-cs2103-ay2324s1.github.io/website/admin/ip-w4.html).
2. Double-click it.
3. Add your tasks.

And the best part is, its **FREE**!! 🤩

## Walk Through
To add `todo`, 
```
todo run
``` 
To add `deadline`, 
```
deadline email /by 11-11-2023 19:00:00
```
To add `event`, 
```
event sports day /from 11-11-2023 08:00:00 /to 12-11-2023 20:00:00
```

## Future Improvements 
- [ ] Improve GUI

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

kayabuttertoastt committed:
Level 0

kayabuttertoastt committed:
Level 1

kayabuttertoastt committed:
Level 2

kayabuttertoastt committed:
Level 3

kayabuttertoastt committed:
Level 4

kayabuttertoastt committed:
A-TextUiTesting

kayabuttertoastt committed:
Level 5

kayabuttertoastt committed:
Level 6

kayabuttertoastt committed:
Add task list save and load functionality from disk

kayabuttertoastt committed:
Improve date handling

kayabuttertoastt committed:
Incorporate OOP

kayabuttertoastt committed:
Add Packages

kayabuttertoastt committed:
Revert "Add Packages"

This reverts commit fffa30d31b613cfbdf32085f74b6674924949ba8.

kayabuttertoastt committed:
Add Packages

kayabuttertoastt committed:
Merge pull request #1 from kayabuttertoastt/add-gradle-support

Add gradle support

kayabuttertoastt committed:
Merge branch 'master' of https://github.com/kayabuttertoastt/ip

* 'master' of https://github.com/kayabuttertoastt/ip:
  Bump gradle and lib version
  Add Gradle support

kayabuttertoastt committed:
Add JUnit tests

kayabuttertoastt committed:
Add JavaDocs

kayabuttertoastt committed:
Edit code to follow the Coding Standard

kayabuttertoastt committed:
Add find feature

xavierpok commented on src/main/java/chatbot/commands/Command.java:
```diff
@@ -0,0 +1,24 @@
+/**
+ * Represents a command that can be executed by the chatbot
+ */
+public abstract class Command {
+    /**
+     * Abstract method that returns a boolean value to indicate whether to exit the program
+     * @return A boolean value
+     */
+    public abstract boolean isExit();
+
+    /**
+     * Abstract method that executes the command that adds the task
+     * @param tasksList Task list which contains the tasks
+     * @param ui A UI instance that displays a message to indicate to the user the task has been added
+     * @param storage Storage instance that represents the storage of the file
+     */
+    public abstract void execute(TaskList tasksList, Ui ui, Storage storage);
+}
```
I like the nice usage of inheritance here, seems to make it much more extensible!

xavierpok commented on src/main/java/chatbot/parser/Parser.java:
```diff
@@ -0,0 +1,75 @@
+     * @return Returns a boolean that indicates if the input is valid
+     * @throws DukeException If the user input is not valid
+     */
+    public static boolean isValid (String command) throws DukeException {
+        if(command.equals("list") || command.equals("bye")) {
+            return true;
+        } else if(command.split(" ")[0].equals("todo") && command.split(" ")[1].length() < 1) {
+            throw new DukeException("Error! Please include a description of the todo");
+        } else if(command.split(" ")[0].equals("deadline") && command.split(" ")[1].length() < 1
+                && !command.contains("/by")) {
+            throw new DukeException("Error! Input should have a description and date");
+        } else if(command.split(" ")[0].equals("event") && command.split(" ")[1].length() < 1
+                && !command.contains("/from") && !command.contains("/to")) {
+            throw new DukeException("Error! Input should have a description, start date and end date");
+        } else if(command.split(" ")[0].equals("mark") && command.split(" ")[1].length() < 1) {
+            throw new DukeException("Error! Please include a description of the task to be marked");
+        } else if(command.split(" ")[0].equals("unmark") && command.split(" ")[1].length() < 1) {
+            throw new DukeException("Error! Please include a description of the task to be unmarked");
+        } else {
+            return true;
+        }
+    }
```
Maybe this could be delegated to the Command classes to figure out what kind of syntax they should accept?

Granted, implementation might get a bit messy, and this certainly works!

xavierpok commented on src/main/java/chatbot/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,54 @@
+ * Represents a command to add a deadline to the task list
+ */
+public class DeadlineCommand extends Command {
+    /**
+     * Description of the deadline to be added
+     */
+    String description;
+
+    /**
+     * Due date for the task
+     */
+    String date;
```
Minor issue, but these fields are lacking access modifiers. Maybe they should be set to private?

```suggestion
    /**
     * Description of the deadline to be added
     */
    private String description;

    /**
     * Due date for the task
     */
    private String date;
```

xavierpok commented on src/main/java/chatbot/commands/EventCommand.java:
```diff
@@ -0,0 +1,61 @@
+ * Represents a command to add an event task to the task list
+ */
+public class EventCommand extends Command {
+    /**
+     * Description of the event task to be added
+     */
+    String description;
+
+    /**
+     * Start date of the event
+     */
+    String from;
+
+    /**
+     * End date of the event
+     */
+    String to;
```
This might need access modifiers to be more explicit
```suggestion
    /**
     * Description of the event task to be added
     */
    private String description;

    /**
     * Start date of the event
     */
    private String from;

    /**
     * End date of the event
     */
   private  String to;
```

xavierpok commented on src/main/java/chatbot/tasks/Event.java:
```diff
@@ -0,0 +1,72 @@
+ * Represents an event
+ */
+public class Event extends Task {
+    /**
+     * End date of event task
+     */
+    protected String end;
+    /**
+     * Start date of event task
+     */
+    protected String start;
+
+    /**
+     * Formatted start date of event task
+     */
+    protected LocalDateTime formattedStart;
+    /**
+     * Formatted end date of event task
+     */
+    protected LocalDateTime formattedEnd;
```
I'm not entirely sure why these are currently `protected`, maybe they could be set to `private`?
```suggestion
    /**
     * End date of event task
     */
    private String end;
    /**
     * Start date of event task
     */
    private String start;

    /**
     * Formatted start date of event task
     */
    private LocalDateTime formattedStart;
    /**
     * Formatted end date of event task
     */
    private LocalDateTime formattedEnd;
```

xavierpok reviewed (commented):
Overall, LGTM, just a few nits.

ItsTYtan commented on src/main/java/chatbot/parser/Parser.java:
```diff
@@ -0,0 +1,75 @@
+     * @return Returns a boolean that indicates if the input is valid
+     * @throws DukeException If the user input is not valid
+     */
+    public static boolean isValid (String command) throws DukeException {
+        if(command.equals("list") || command.equals("bye")) {
+            return true;
+        } else if(command.split(" ")[0].equals("todo") && command.split(" ")[1].length() < 1) {
+            throw new DukeException("Error! Please include a description of the todo");
+        } else if(command.split(" ")[0].equals("deadline") && command.split(" ")[1].length() < 1
+                && !command.contains("/by")) {
+            throw new DukeException("Error! Input should have a description and date");
+        } else if(command.split(" ")[0].equals("event") && command.split(" ")[1].length() < 1
+                && !command.contains("/from") && !command.contains("/to")) {
+            throw new DukeException("Error! Input should have a description, start date and end date");
+        } else if(command.split(" ")[0].equals("mark") && command.split(" ")[1].length() < 1) {
+            throw new DukeException("Error! Please include a description of the task to be marked");
+        } else if(command.split(" ")[0].equals("unmark") && command.split(" ")[1].length() < 1) {
+            throw new DukeException("Error! Please include a description of the task to be unmarked");
+        } else {
+            return true;
+        }
```
Could perhaps use a switch statement and split the string command at the top, since some splits overlap in the code. I think the validation of the command can be done at the same time as parsing the command and deciding which command to execute.

ItsTYtan commented on src/main/java/chatbot/tasks/Deadline.java:
```diff
@@ -0,0 +1,57 @@
+     * @return Formatted string representation of the deadline
+     */
+    @Override
+    public String store() {
+        return String.format("D | %s | %s | %s", this.isDone, this.description,
+                formattedDate.format(DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm:ss")));
+    }
```
Nice use of formatters

ItsTYtan commented on src/main/java/chatbot/tasks/TaskList.java:
```diff
@@ -0,0 +1,103 @@
+     * Removes the task at a particular index as done
+     * @param num Index of the particular task in the task list
+     */
+    public void removeTask(int num) {
+        if(num < 0){
+            System.out.println("The value cannot be negative. Please try again.");
+        } else if (num >= this.inputs.size()){
+            System.out.println("The input value is wrong. Please try again.");
+        } else {
+            this.inputs.remove(num);
+        }
+    }
```
Maybe have an exception for invalid nums?

ItsTYtan commented on src/main/java/chatbot/tasks/TaskList.java:
```diff
@@ -0,0 +1,103 @@
+     * Retrieves the list of tasks stored in the task list
+     * @return ArrayList that contains the tasks stored in the task list
+     */
+    public ArrayList<Task> retrieveList() {
+        return this.inputs;
+    }
```
Not sure if its a good idea to retrieve the arraylist in the tasklist class itself. I think any operation on the tasklist can be done without exposing the inner mutatable array

ItsTYtan reviewed (commented):
LGTM, just some nitpicking on minor issues.

kayabuttertoastt committed:
Set up Checkstyle

nicrandomlee commented on src/main/java/chatbot/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    /**
+     *  Returns a boolean value to indicate whether to exit the program
```
Extra spacing in javadoc

nicrandomlee commented on src/main/java/chatbot/parser/Parser.java:
```diff
@@ -0,0 +1,84 @@
+
+        isValid(command);
+
+        if (command.equals("bye")) {
```
You can try doing a switch statement to make the code more readable. Feel free to use enums to represent commands as well so it is clearer.

Eg:
case BYE:
<insert code here>

nicrandomlee commented on src/main/java/chatbot/parser/Parser.java:
```diff
@@ -0,0 +1,84 @@
+        if (command.equals("list") || command.equals("bye")) {
+            return true;
+        } else if (command.split(" ")[0].equals("todo") && command.split(" ")[1].length() < 1) {
+            throw new DukeException("Error! Please include a description of the todo");
+        } else if (command.split(" ")[0].equals("deadline") && command.split(" ")[1].length() < 1
+                && !command.contains("/by")) {
+            throw new DukeException("Error! Input should have a description and date");
+        } else if (command.split(" ")[0].equals("event") && command.split(" ")[1].length() < 1
+                && !command.contains("/from") && !command.contains("/to")) {
+            throw new DukeException("Error! Input should have a description, start date and end date");
+        } else if (command.split(" ")[0].equals("mark") && command.split(" ")[1].length() < 1) {
+            throw new DukeException("Error! Please include a description of the task to be marked");
+        } else if (command.split(" ")[0].equals("unmark") && command.split(" ")[1].length() < 1) {
+            throw new DukeException("Error! Please include a description of the task to be unmarked");
+        } else {
```
Just nitpicking a little but you can consider adding the format that the user is supposed to enter so that they have a point of reference, since most of the error messages are generic for the particular task entered and the user may not know what went wrong.

nicrandomlee commented on src/main/java/chatbot/tasks/TaskList.java:
```diff
@@ -0,0 +1,102 @@
+     */
+    public void removeTask(int num) {
+        if (num < 0) {
+            System.out.println("The value cannot be negative. Please try again.");
```
Good handling of edge cases!

nicrandomlee reviewed (commented):
Overall, I found your code really easy to read. Good use of OOP to handle various commands and coding standards too! I am very thankful to have reviewed your code as I learnt a lot from your code writing style. Keep it up!

Cleon2 commented on src/main/java/chatbot/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,54 @@
+ * Represents a command to add a deadline to the task list
+ */
+public class DeadlineCommand extends Command {
+    /**
```
Need to add spacing here for the javadoc

Cleon2 commented on src/main/java/chatbot/commands/TodoCommand.java:
```diff
@@ -0,0 +1,48 @@
+     * @param description Description of the todo task to be added
+     */
+    public TodoCommand(String description) {
+
```
unnecessary spacing here

Cleon2 commented on src/main/java/chatbot/commands/Command.java:
```diff
@@ -0,0 +1,24 @@
+/**
+ * Represents a command that can be executed by the chatbot
+ */
+public abstract class Command {
```
Love how you abstracted everything using command!

Cleon2 commented on src/main/java/chatbot/parser/Parser.java:
```diff
@@ -0,0 +1,84 @@
+     * @return Returns a boolean that indicates if the input is valid
+     * @throws DukeException If the user input is not valid
+     */
+    public static boolean isValid (String command) throws DukeException {
```
Code seems a bit wordy here

Cleon2 reviewed (commented):
Overall looks good o

kayabuttertoastt committed:
Add GUI

kayabuttertoastt committed:
Add assertions

Command classes: Assertions were not implemented.

This caused key assumptions to not be held at various
points in the code.

Assertions were used to ensure these key assumptions are held.

kayabuttertoastt committed:
Merge pull request #3 from kayabuttertoastt/branch-Assertions

Add assertions

kayabuttertoastt committed:
Improve code quality

Code had minor code quality issues.

Issues resolved:
* Removed dead code
* Edited header comments to follow specified format
* Changed access modifiers
* Edited error messages to be more specific

kayabuttertoastt committed:
Merge pull request #4 from kayabuttertoastt/branch-CodeQuality

Improve code quality

kayabuttertoastt committed:
Add Help Command

Provide in-App guidance to users by adding a help command that can be executed by the user when they need help.

Invalid Command also contains valid commands that the user can execute.

kayabuttertoastt committed:
Merge pull request #5 from kayabuttertoastt/branch-C-Help

Add Help Command

kayabuttertoastt committed:
Improve code quality

kayabuttertoastt committed:
Edit error messages

Edited error messages to be more specific and fixed bugs related to the display of messages when an exception is thrown.

kayabuttertoastt committed:
Improve GUI

Added welcome message and spacing in between responses for better UI.

kayabuttertoastt committed:
Improve GUI

* Added padding between responses as well as between image and text
* Changed alignment such that chatbot responses are left-aligned and user responses are right-aligned.

kayabuttertoastt committed:
Add Ui.png

kayabuttertoastt committed:
Delete Ui.png

kayabuttertoastt committed:
Add Ui.png

kayabuttertoastt committed:
Add Javadocs and Rename Application

Changes:
* Application is renamed from "Chatbot" to "Gudetama" to make the images in the GUI
* Javadocs is added to most of the classes and their methods

kayabuttertoastt committed:
Modify README.md

Modifications:
* Include brief introduction to the application
* Include Features of the application
* Include Usage of features

kayabuttertoastt committed:
Added more testcases

Added testcase to check HelpCommand and fixed bugs.

kayabuttertoastt committed:
Fixed Gradle Bugs

kayabuttertoastt committed:
Fixed bugs and error messages

Fixed bugs related to duke.txt file and added more specific error messages for mark and unmark commands

kayabuttertoastt committed:
Fixed bugs

Fixed bugs related to creation of duke.txt file

kayabuttertoastt committed:
Fixed bugs and updated javadocs

Fixed storage bugs and updated javadocs

kayabuttertoastt committed:
Updated goodbye message

Updated goodbye message and regenerated duke.jar file
